testng-core/src/main/java/org/testng/SuiteRunner.java

```diff
@@ -262,14 +236,17 @@ public ITestListener getExitCodeListener() {
   }
 
   private void invokeListeners(boolean start) {
+    List<ISuiteListener> original = Lists.newArrayList(listeners.values());
+    ListenerComparator comparator = this.configuration.getListenerComparator();
+    if (comparator != null) {
+      original.sort(comparator::compare);
+    }
     if (start) {
-      for (ISuiteListener sl :
-          ListenerOrderDeterminer.order(Lists.newArrayList(listeners.values()))) {
+      for (ISuiteListener sl : ListenerOrderDeterminer.order(original)) {
```

Comment:
`ListenerOrderDeterminer` should be replaced by the `ListenerComparator`.
Maybe `ListenerOrderDeterminer` should be the default `ListenerComparator`.

Reply:
Are you suggesting that we move all the `static` methods within `ListenerOrderDeterminer` into `ListenerComparator`? 

`ListenerOrderDeterminer` is more of like a specialised sorting utility that iterates through a given list and seggregates them into normal listeners and special listeners (these are listeners that are provided via the JVM argument `-Dtestng.preferential.listeners.package` with a default value of `"com.intellij.rt.*"`) and then orders them (either in original order or reversed order) such that the normal listeners come first followed by the special listeners.

But yes, I can move all the static methods into just `ListenerComparator` so that it's easy to find all the listener ordering logic in one place.

Reply:
fix it by moving the sort into ListenerOrderDeterminer